platform/core/PolyPodCoreAndroid/core/src/main/kotlin/coop/polypoly/core/MessagePackDecoding.kt

```diff
@@ -34,33 +39,54 @@ fun mapError(msgObject: Map<Value, Value>): CoreFailure {
     throw InvalidCoreErrorFormat(msgObject.toString())
 }
 
-fun mapFeatureManifest(msgObject: Value): FeatureManifest {
-    val msgObject = msgObject.asMapValue().map() ?: throw EmptyFeatureManifest()
+fun mapFeatureCategories(msgObject: Value): List<FeatureCategory> {
+    return msgObject.asArrayValue().map {
+        mapFeatureCategory(it)
+    }
+}
+
+fun mapFeatureCategory(msgObject: Value): FeatureCategory {
+    val map = msgObject.asMapValue().map()
+
+    return FeatureCategory(
+        FeatureCategoryId.valueOf(map.getValue("id").toString()),
+        name = map.getValue("name").toString(),
+        features = mapFeatures(map.getValue("features"))
+    )
+}
+
+fun mapFeatures(msgObject: Value): List<Feature> {
+    return msgObject.asArrayValue().map {
+        mapFeature(it)
+    }
+}
 
-    val name = msgObject.get("name")?.getStringValue()
-    val author = msgObject.get("author")?.getStringValue()
-    val version = msgObject.get("version")?.getStringValue()
-    val description = msgObject.get("description")?.getStringValue()
-    val thumbnail = msgObject.get("thumbnail")?.getStringValue()
-    val thumbnailColor = msgObject.get("thumbnailColor")?.getStringValue()
-    val primaryColor = msgObject.get("primaryColor")?.getStringValue()
-    val borderColor = msgObject.get("borderColor")?.getStringValue()
-    val tileTextColor = msgObject.get("tileTextColor")?.getStringValue()
-    val links = msgObject.get("links")?.getMapValue()
+fun mapFeature(msgObject: Value): Feature {
+    val map = msgObject.asMapValue().map()
 
-    return FeatureManifest(
-        name,
-        author,
-        version,
-        description,
-        thumbnail,
-        thumbnailColor,
-        primaryColor,
-        borderColor,
-        tileTextColor,
-        links?.map {
+    return Feature(
+        map.getValue("path").toString(),
+        id = map.getValue("id").toString(),
+        name = map.getValue("name").toString(),
+        author = map.get("author")?.toString(),
+        version = map.get("version")?.toString(),
+        description = map.get("description")?.toString(),
+        thumbnail = map.get("thumbnail")?.toString(),
+        thumbnailColor = Color.parseColor(
```

Comment:
Seeing how many colors we have, and the boilerplate they generate, could this be a map instead of variables with different values? That way we could loop over the manifest to assign values.

Reply:
Those are meant to be statically checked and assures the fields existence, using a map would make this dynamic without any compile time checks. The boilerplate is caused by doing manual MessagePack decoding(which ideally should go away), not by how colours are defined.